ql/src/java/org/apache/hadoop/hive/ql/Driver.java

```diff
@@ -213,7 +216,14 @@ private void runInternal(String command, boolean alreadyCompiled) throws Command
         releaseResources();
       }
 
+      if (SessionState.get() != null) {
+        // Clean up every resource object stored in the query state
+        driverContext.getQueryState().removeResources();
```

Comment:
Do we need this one?
We remove the `QueryState` anyway...
And then we do not need `removeResources` either

Reply:
I think we need it. We have an instance of `QueryState` stored on the `DriverContext`, and we have the same `QueryState` instance stored on `SessionState` as well. If I remove the `QueryState` just from the `SessionState`, we will end up with some dangling resources on the `QueryState` bound to the `DriverContext`.

Reply:
The QueryState in the DriverContext should be removed as well when the query finishes. Otherwise we still would have a leaked QueryState which is a memory leak itself. So this only makes the leak smaller, but does not fix it (if we have any)

Reply:
I've deleted the `removeResources`.